chebi-update/src/main/java/org/reactome/release/chebiupdate/ChebiUpdater.java

```diff
@@ -395,39 +535,75 @@ private void retrieveUpdatesFromChebi(Collection<GKInstance> refMolecules, Map<L
-					if (!identifier.startsWith("CHEBI:"))
+					Entity entity;
+					// only query web service if the data is not in the chebi-cache - NOTE: chebiCache will always be empty if this.useCache == false
+					if (!chebiCache.containsKey("CHEBI:"+identifier))
 					{
-						identifier = "CHEBI:" + identifier;
+						if (this.useCache && chebiCache.size() > 0)
+						{
+							logger.trace("Cache miss for CHEBI:{}", identifier);
+						}
+						entity = this.chebiClient.getCompleteEntity(identifier);
+						if (entity != null && this.useCache)
+						{
+							bw.write("CHEBI:"+identifier+"\t"+entity.getChebiId()+"\t"+entity.getChebiAsciiName()+"\t"+ (entity.getFormulae().size() > 0 ? entity.getFormulae().get(0).getData() : "") + "\t" + LocalDateTime.now().toString() + "\n");
+							bw.flush();
+						}
+						else
+						{
+							if (entity == null)
+							{
+								failedEntitiesList.put(molecule, "ChEBI WebService response was NULL.");
+							}
+						}
 					}
-					Entity entity = this.chebiClient.getCompleteEntity(identifier);
-					if (entity != null)
+					else // ...Load data from the cache. Use the AccessibleEntity to set the formula.
 					{
-						entityMap.put(molecule.getDBID(), entity);
+						entity = new AccessibleEntity();
+						entity.setChebiId(chebiCache.get("CHEBI:"+identifier).get(0));
+						entity.setChebiAsciiName(chebiCache.get("CHEBI:"+identifier).get(1) );
+						DataItem formula = new DataItem();
+						formula.setData(chebiCache.get("CHEBI:"+identifier).get(2));
+						((AccessibleEntity)entity).setFormulae(Arrays.asList(formula));
 					}
-					else
+					// Add entity to map (if it's non-null)
+					if (entity != null)
 					{
-						failedEntitiesList.add(molecule);
+						entityMap.put(molecule.getDBID(), entity);
 					}
 				}
+				else
+				{
+					logger.error("ERROR: Instance \"{}\" has an empty/null identifier. This should not be allowed.", molecule.toString());
+					failedEntitiesList.put(molecule, molecule.toString() + " has an empty/NULL identifier.");
+				}
+				int i = counter.getAndIncrement();
+				if (i % 250 == 0)
+				{
+					logger.debug("{} ChEBI identifiers checked", i);
+				}
 			}
 			catch (ChebiWebServiceFault_Exception e)
 			{
-				try
+				// "invalid ChEBI identifier" probably shouldn't break execution but should be logged for further investigation.
+				if (e.getMessage().contains("invalid ChEBI identifier"))
 				{
-					logger.error("WebService error: {} ; Identifier that triggered it: {}; Molecule DB ID: {}", e.getMessage(), identifier, molecule.getDBID());
+					logger.error("ERROR: ChEBI Identifier \"{}\" is not formatted correctly.", identifier);
+					failedEntitiesList.put(molecule, "ChEBI Identifier \""+identifier+"\" is not formatted correctly.");
 				}
-				catch (Exception e1)
+				// Log this identifier, but don't fail.
+				else if (e.getMessage().contains("the entity in question is deleted, obsolete, or not yet released"))
```

Comment:
Maybe using .matches with keywords and "OR" in a regex would be better in case ChEBI changes their error message slightly

Reply:
Do you mean something like
```
...matches(".*the |entity |in |question |is |deleted....*");
```
Anticipating possible future changes to messages seems iffy. If ChEBI sends a `ChebiWebServiceFault_Exception` that has a different message, the application will still log it.

Reply:
Could it look for just for 'deleted', 'obsolete', or 'not yet released'?  My concern was that if the error message doesn't match exactly, the else clause runs and a RuntimException is triggered.

Reply:
There could be other error messages that contain those words but don't mean the same thing as the message I'm checking for now. In this case I'd rather let the program fail on a completely new error message, than let it continue running when it finds a completely new message that might need to be handled in a new way.

Reply:
Okay, that makes sense